Add window functions design doc

ggevay opened the pull request:
This PR moves [the window functions design doc from Notion](https://www.notion.so/materialize/Efficient-Window-Functions-87682d6de73a4f6088acde21aba41fe6) to a markdown file.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,251 @@
+
+### Recognizing the current windowing pattern in just the rendering
+
+There is one more option that bypasses any MIR/LIR modifications: we could leave HIR, MIR, and LIR unchanged, and just make the rendering detect that pattern that the current HIR lowering creates, and do a custom rendering to it. However, this might be too brittle: we might accidentally leave out cases when the pattern somehow ends up slightly different, plus we might break it from time to time w ...
```
Moving an open comment thread from Notion to here:

Frank:
I’m not sure I follow this argument. I’m pretty sure that we can (perhaps already do) have a scalar variant for window functions that we 100% have no clue how to transform, which ensures that we won’t accidentally remove it or transform it or anything.

ggevay replied:
The current HIR lowering creates this entire big thing that the example here in the gray box shows:
- a Reduce with a 3-level nested row, 
- a FlatMap with an unnest_list,
- a Map with a bunch of record_gets,
- and some Projects around these.

This pattern does have a `lag(...)`, which is indeed a definitive indication that there is windowing going on. It’s also true that this lag definitely won’t get transformed away. However, to introduce a custom rendering based on Prefix Sum, it’s not enough to pattern-match just this lag, but we need to match the entire pattern of Reduce-Project-FlatMap-Map-Project.

One example complication in matching this pattern, is that MFP CSE can make the expressions of the Map look different from what the HIR lowering creates:
- If there is only one field being projected (besides the window function) at the end, then we will have a `record_get[xxx](record_get[1](#1))`
- But if we are projecting multiple fields, then MFP CSE will create a `record_get[1](#1)`, which goes into `#2`, and then there will be a `record_get[xxx](#2)` for each of the fields. It’s not impossible to match both of these cases, it’s just a bit cumbersome and possibly error-prone.

ggevay replied:
In the meantime, we had a discussion in the office hours, and a new proposal was to create a many-to-many Reduce as a new MirRelationExpr enum variant. This would initially only contain windowing, but after the windowing is done, we would probably merge TopK into this.

frankmcsherry replied:
To loop back around, one imo totally valid proposal is to change nothing about the IR and use idiom detection in rendering, or potentially just in LIR. We already do this in several cases, most obviously with MFPs around operators that can be taught to absorb them (e.g. an MFP around a FlatMap, as in these examples).

The advantage this has is that you have already written correct implementations of projection pushdown, demand propagation, non-nullable requirements pushdown, the `typ()` function along with key detection and nullability of columns. All of this needs to be written anew for each variant that is a black box unrelated to the existing primitives. 

Looking at the code snippet below, the only defect seems to be that we don't absorb `Project` into `FlatMap`, which we should do in any case. At this point the idiom is "`FlatMap` with `unnest_list` atop `Reduce` with certain aggregates".

My understanding is that the specific reason this doesn't work for you is that someone needs to name a specific type for the ordering key, and the types are no longer visible in LIR if they are not captured. We could plausibly capture them in `lag` .. or each of the functions that supports the prefix sum representation.

My understanding is that nonetheless you would prefer to introduce new variants, but, as one of the people who needs to help fix all of the existing transforms, my default disposition is that you not introduce new variants, as it is more code to maintain.

ggevay replied:
Answering this [here](https://github.com/MaterializeInc/materialize/pull/17077#issuecomment-1403712012).

antiguru commented:
This looks great! I'm not yet hitting approve as I need to ponder on what steps this should improve. @frankmcsherry, if you're happy with the contents, can you LGTM? Thanks!

ggevay commented:
@antiguru The MIR and LIR representations are still open:
- MIR: I think [the top option now for MIR](https://github.com/MaterializeInc/materialize/pull/17077/files#r1067209676) is kind of a hybrid between 1. and 3.: create a many-to-many Reduce as a new `MirRelationExpr` enum variant. This variant would initially only contain windowing, but after the windowing is done, we would probably merge TopK into this.
- LIR: We haven't discussed much with @frankmcsherry about this yet, but it seems to me that a new variant would be the best option, because the rendering of the current `Reduce` or `TopK` wouldn't share any code with the rendering of windowing. (But in any case, this seems like a less critical question as MIR, since we don't have LIR transforms.)

I'm waiting on feedback from @frankmcsherry about these questions, and then I'll include this in the design doc before merging.

frankmcsherry commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+We’ll use the word ****index**** in the below text to mean the values of the ORDER BY column of the OVER clause, i.e., they are simply the values that determine the ordering. (Note that it’s sparse indexing, i.e., not every number occurs from 1 to n, but there are usually (big) gaps.)
+
+Now I’ll list all window functions, and how we’ll support them with one of the above approaches:
```
In reading this, it starts as an enumeration of window functions, but turns in to a bullet salad. If at all possible, it would help to decouple the moments that are meant to be enumerations from the associated thoughts, to make it easier for the reader to assess the exhaustiveness / find the special case they are looking to understand.

ggevay replied:
Yes, sorry, I didn't realize how annoying can a bullet salad be. I completely changed the structuring. The "How to handle each window function" section has numbered subsections, enumerating all window functions (grouped into categories). There are also two big horizontal lines delineating this big enumeration section.

frankmcsherry commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+(And this is just the MIR, but we would work with the LIR.)
+
+Also, we will want other transforms for window functions, most notably, the ROW_NUMBER-to-TopK transform. For this, we need to canonicalize scalar expressions, which I think we usually do in MIR. This means that this transform should happen on MIR, meaning that we would need an MIR representation of window functions.
```
I struggle to understand your use of "need" here. It seems not hard to write a transform that looks for the expected idiom of a filter over the `row_number` window function block. That transform may not be great, it may be brittle, but it is a thing we could do. 

Is there a technical **requirement** to have an MIR variant for window functions, or is it just something that will make window functions themselves easier to implement and maintain (at the expense of the rest of the codebase)?

ggevay replied:
> Is there a technical **requirement** to have an MIR variant for window functions

Ok, there is no requirement. I changed the design doc to go with the idiom recognition approach also for the TopK pattern.

frankmcsherry commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+    - Compared to 3., it might be easier to skip windowing in many transforms. This is both good and bad:
+        - We can get a first version done more quickly. (And then potentially add optimizations later.)
+        - But we might leave some easy optimization opportunities on the table, which would come from already-existing transform code for `Reduce`.
+   - A new `MirRelationExpr` variant would mean we have to modify about 12-14 transforms (`LetRec` is pattern-matched in 12 files in the `transform` crate, `TopK` 14 times. See also [the recent LetRec addition](https://github.com/MaterializeInc/materialize/commit/9ac8e060d82487752ba28c42f7b146ff9f730ca3) for an example of how it looks when we add a new `MirRelationExpr` variant.)
```
For clarity, `LetRec` is disabled in all but a few transforms because we don't know that it is correct without a more thorough audit of the transforms. We have non-exhaustive matches (e.g. predicate pushdown) where I suspect it is not correct, and .. we'll have to do the same sort of checking for a `WindowFunction` variant).

ggevay replied:
I changed the text to clarify this.

frankmcsherry reviewed (commented):
I got this far before I need a break. One thing that is challenging is that the format is not prose but rather bulleted lists. I have a hard time reading this and trying to assess what points are being made, whether they are clearly made, whether we are moving on to the next point or musing on something else. I do understand what you want to do, and we are getting to the point where I'll say "yes" just to not have to fight it, but I think there is a legit exercise in recording the reasoning that we want to do before we move forward.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+**Duplicate indexes.** There might be rows which agree on both the PARTITION BY key and the ORDER BY key (the index). Prefix Sum doesn’t handle such duplicate indexes.
+
+- To eliminate these duplicates, we will number the elements inside each group with 0, 1, 2, …, and this will be an additional component of the prefix sum indexes.
```
The problem with that is that the entire `Row` might match. Note that in other ordering situations this possibility is not a problem, because if the entire `Row` matches, then there is no difference between ordering them one way or the other. But the issue here is not ordering stability, but that the current Prefix Sum implementation just cannot handle duplicate ordering keys.

I added a note that an alternative would be to make Prefix Sum correctly handle duplicate ordering keys.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+            - It doesn’t have a zero element if implemented naively, but we can add one artificially: Normally we always return the right argument, but if the right argument is the artificially added zero element, then we return the left argument. Note: adding the zero element is important, because prefix sum sometimes sums in the zero element not just at the beginning, but randomly in the middle ...
+            - I built [a small prototype outside Materialize](https://github.com/ggevay/window-funcs), where I verified that the output values are correct, and that output is quickly updated for small input changes.
+        - For LAG/LEAD with offset *k > 1* (which computes the given expression not for the previous record, but for the record that was *k* records ago):
+            - A sum function could simply remember the last *k* values. This works kind of ok for small *k*.
```
Thanks, I updated the sentence.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+            - I built [a small prototype outside Materialize](https://github.com/ggevay/window-funcs), where I verified that the output values are correct, and that output is quickly updated for small input changes.
+        - For LAG/LEAD with offset *k > 1* (which computes the given expression not for the previous record, but for the record that was *k* records ago):
+            - A sum function could simply remember the last *k* values. This works kind of ok for small *k*.
+                - There is a technical trick to make the overhead smaller by avoiding doing many copies on the list that is keeping the last **k** values: We could use such a reduction that has a `&mut` argument, and keep the last **k** values in that argument.
```
The *k* and **k** difference came about somehow by accident when copy-pasting from Notion.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+        - For LAG/LEAD with an offset of 1, the sum function will just remember the previous value. (And we can have a similar one for LEAD.) This looks a bit weird as a sum function, but it will work:
+            - it's associative;
+            - It’s not commutative, but that shouldn’t be an issue for Prefix Sum;
+            - It doesn’t have a zero element if implemented naively, but we can add one artificially: Normally we always return the right argument, but if the right argument is the artificially added zero element, then we return the left argument. Note: adding the zero element is important, because prefix sum sometimes sums in the zero element not just at the beginning, but randomly in the middle ...
```
Yes, thanks, simplified the text.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+Now I’ll list all window functions, and how we’ll support them with one of the above approaches:
+
+- Frameless window functions
```
Added short explanations and links to Postgres docs.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+1. We’ll use [DD’s prefix_sum](https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/algorithms/prefix_sum.rs) with some tricky sum functions.
+2. As an extension of 1., we’ll use a generalization of DD’s prefix sum to arbitrary intervals (i.e., not just prefixes).
+3. We transform away the windowing in some special cases (e.g., to TopK, or a simple grouped aggregation + self-join)
```
Yes, thanks, I added resorting to the old window function implementation as a 4. case.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,255 @@
+
+## Current state
+
+The current way of execution is to put entire partitions into scalars, and execute the window function to all elements by a “scalar aggregation”. This happens in the HIR to MIR lowering, i.e., MIR and LIR don’t know about windowing.
```
Done

ggevay commented:
Thank you very much for the comments, @frankmcsherry! I addressed all of them.

> To loop back around, one imo totally valid proposal is to change nothing about the IR and use idiom detection in rendering, or potentially just in LIR.

Yes. I've changed the design doc to go with the idiom recognition approach.

> Looking at the code snippet below, the only defect seems to be that we don't absorb Project into FlatMap, which we should do in any case.

Sorry, I don't really understand this. So we have
```
   Project (#3..=#5)                                                                             +
     Map (record_get[1](#1), record_get[0](#2), record_get[2](#2), record_get[0](#1))            +
       FlatMap unnest_list(#0)                                                                   +
         Project (#1)       
```
Do you mean that we should change the `FlatMap` to refer to `#1` instead of `#0`? But then we'll still have the extra column after the `FlatMap` is done, because `FlatMap` _appends_ a column (similarly to `Map`). I don't see how we could make the `FlatMap` perform also the projection. Maybe the `Project` after the `Map` could project away the extra column, but then I wouldn't call this "absorbing the Project", but it would be something like projection lifting.

> My understanding is that the specific reason this doesn't work for you is that someone needs to name a specific type for the ordering key, and the types are no longer visible in LIR if they are not captured. 

In the meantime, I'm thinking to do the idiom recognition in the MIR-to-LIR lowering, and then there is no type problem. (Also, AFAIU we shouldn't have conditional code in the rendering, and this idiom recognition will be a giant piece of conditional code, so it seems to belong more in the lowering. Also, we want EXPLAIN PHYSICAL PLAN to show how we'll execute a window function.)

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+[Many users want to use window functions](https://www.notion.so/Window-Functions-Use-Cases-6ad1846a7da942dc8fa28997d9c220dd), but our current window function support is very inefficient: We recompute results for an entire window partition for any small change in the partition. So the only situations when our current support works is if the window partitions are either very small, or they rarely c ...
+
+## Goals
```
This section seems to mix goals with implementation, for example that some things are easy and others are not, or that `prefix_sum` will handle things for us. Ideally, this section lays out what we'll have after applying the change to Materialize that we didn't have before. The actual mapping should be part of the design section I believe.

ggevay replied:
I mostly separated them, but I think a few short hints to which stuff will be supported by prefix sum vs. translating away window functions is still fine to have here.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+Some window functions are impossible to efficiently support in streaming, because sometimes small input changes cause big result changes. (E.g., if a new first element of a partition appears, then ROW_NUMBERs will change for the whole window partition.) So a realistic goal would be to support at least those cases where a small input change leads to a small output change.
+
+- LAG/LEAD (i.e., previous/next element of the window partition)
+    - We aim for only offset 1 in the first version, which is the default. Bigger offsets not seen in user queries yet, but shouldn’t be a problem to add support later.
```
"shouldn’t be a problem to add support later." -> Please elaborate, it's not clear why that is, or why we don't do it now.

ggevay replied:
Moving this to the implementation section.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+    - TopK is an important special case: This is when there is a `WHERE ROW_NUMBER() < k`
+        - Instead of doing Prefix Sum, we will transform this into our [efficient TopK implementation](https://www.notion.so/e62fe2b3d8354052ac7d0fe92be1e711).
+
+## Non-goals / Limitations
```
I personally would not mix non-goals and limitations. Non-goals are things that you explicitly don't want to do, but limitations are part of the discussion of your solution.

ggevay replied:
Hmm, yes. I separated them into two different sections.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+For invertible aggregation functions (e.g., sum, but not min/max) we can use the existing prefix sum with a minor trick: agg(a,b) = agg(0,b) - agg(0,a).
+    - However, the performance of this might not be good, because even if (a,b) is a small interval, the (0,a) and the (0,b) intervals will be big, so there will be many changes of the aggregates of these even for small input changes.
+
+To have better performance (and to support non-invertible aggregations, e.g., min/max), we need to extend what the `broadcast` part of prefix sum is doing (`aggregate` can stay the same):
+    - `queries` will contain intervals specified by two indexes.
+    - `requests`: We can similarly compute a set of requests from `queries`. The change will only be inside the `flat_map`.
+    - `full_ranges`, `zero_ranges`, `used_ranges` stay the same.
+    - `init_states` won’t start at position 0, but at the lower end of the intervals in `queries`
+    - The iteration at the end will be mostly the same.
```
This is specific to to implementation and doesn't make sense for someone who didn't study it. I think you need an explanation of the `prefix_sum` operator that you suggest we use.

ggevay replied:
I added an explanation for prefix_sum's implementation.

ggevay replied:
I will also improve this explanation a bit.

antiguru reviewed (commented):
I didn't finish reading, but I wanted to leave some early feedback. I think this is on track to becoming a good design doc, but there's work left to make the content more approachable. I'll continue tomorrow!

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+### Parallelism
+
+DD's Prefix Sum should be data-parallel even inside a window partition. (It’s similar to [a Fenwick tree](https://en.wikipedia.org/wiki/Fenwick_tree), with sums maintained over power-of-2 sized intervals, from which you can compute a prefix sum by putting together LogN intervals.) TODO: But I wasn't able to actually observe a speedup in a simple test when adding cores, so we should investigate wh ...
```
This sounds like you don't want to make it part of the design, but I strongly suggest that any implementation includes a performance evaluation, and that we outline expected performance here.

ggevay replied:
Ok, yes, I added much more details about performance. Will add more when I have a prototype.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+DD's Prefix Sum should be data-parallel even inside a window partition. (It’s similar to [a Fenwick tree](https://en.wikipedia.org/wiki/Fenwick_tree), with sums maintained over power-of-2 sized intervals, from which you can compute a prefix sum by putting together LogN intervals.) TODO: But I wasn't able to actually observe a speedup in a simple test when adding cores, so we should investigate wh ...
+
+### Types
+
+We'll have to generalize DD's Prefix Sum to orderings over types other than a single unsigned integer, which is currently hardcoded in the code that forms the intervals. We’ll map other types to a single unsigned integer. Importantly, this mapping should *preserve the ordering* of the type:
+
+- Signed integer types are fine, we just need to fiddle with the sign to map them to an unsigned int in a way that preserves the ordering.
+- Date/Time types are just a few integers. We’ll concatenate their bits.
+- I don’t know how to handle strings, so these are out of scope for now. (Not seen in user queries yet.)
```
Can you explain what the requirements for types in prefix sum are? I have a hard time following what types work and which types don't.

ggevay replied:
Yes, I rewrote the entire section.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+    - (I already tested a [simple prototype for LAG outside Materialize](https://github.com/ggevay/window-funcs).)
+- We should measure the memory needs of Prefix Sum, so that we can advise users when sizing replicas.
+
+## Lifecycle
+
+There are many window functions, and many frame options. We will gradually add the new, efficient implementations for an increasing set of window function + frame setting combinations across several PRs. We will make sure the new implementation is correct and performs well before merging each of the PRs.
```
This should be made explicit so we can observe progress.

ggevay replied:
I added a link to the [overall tracking issue](https://github.com/MaterializeInc/materialize/issues/16367).

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+# Open questions
+
+- How to have automated performance tests? Can we check in Testdrive that some materialized view (that has window functions) is being updated fast enough?
```
Since we don't have automated performance tests, I would not block on it. However, we need manual performance tests to validate this design.

ggevay replied:
Ok, I clarified that it's ok to only have manual performance tests for the first version.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+We are [in the process of adding support for recursive SQL queries](https://github.com/MaterializeInc/materialize/issues/17012). We could implement prefix sum using that instead of custom rendering code. In this case, all the window function handling could happen in the HIR-to-MIR lowering, by translating to a MIR expression that has `LetRec`.
+
+### Representing window functions in each of the IRs
```
This section discusses what options we have, but it doesn't discuss trade-offs. It's not a high-priority item, but I think that as we're working on this, we should make an explicit decision which representation we select and for what reasons.

ggevay replied:
I added more discussion on trade-offs of the different representations.

antiguru commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+In most situations other than TopK, these functions cannot be implemented efficiently in a streaming setting, because small input changes often lead to big output changes. However, as noted in the [Goals](#Goals) section, there are some special cases where small input changes will lead to small output changes. These will be possible to support efficiently by performing a Prefix Sum with an approp ...
+
+#### 2. Window aggregations
```
I find this part difficult to follow, but with some cleanup and an explanation of prefix sum it should get better.

antiguru reviewed (commented):
I added more comments. I really think we should start from a problem statement, a discussion of prefix sum, which combined can then result in a solution to said problem. My goal of this effort is to make sure that we consider alternatives and discuss the design such that we avoid surprises on the way, and that we have a reference that allows others to pick up this work at a later point in time. It doesn't need to be the documentation, because design docs are a point-in-time artifact, but it should enable us to get there!

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+**(Note that [temporal windows](https://materialize.com/docs/sql/patterns/temporal-filters/) are entirely different from what we discuss in this doc. Our support for those is already quite fine. There is no need to use SQL’s `OVER` clause for those.)**
+
+# Overview
```
I added an intro about window functions.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+- Date/Time types are just a few integers. We’ll concatenate their bits.
+- I don’t know how to handle strings, so these are out of scope for now. (Not seen in user queries yet.)
+
+## Alternatives
```
Thanks, added them!

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+### Use WITH MUTUALLY RECURSIVE
+
+We are [in the process of adding support for recursive SQL queries](https://github.com/MaterializeInc/materialize/issues/17012). We could implement prefix sum using that instead of custom rendering code. In this case, all the window function handling could happen in the HIR-to-MIR lowering, by translating to a MIR expression that has `LetRec`.
```
I added 2 paragraphs on this, plus a list of pro and contra arguments.

ggevay commented on doc/developer/design/20230110_window_functions.md:
````diff
@@ -0,0 +1,539 @@
+             Get materialize.public.cities                                                       +
+```
+
+To avoid creating a new enum variant in MirRelationExpr, we will recognize the above pattern during the MIR-to-LIR lowering, and create a new LIR enum variant for window functions. I estimate this pattern recognition to need about 15-20 if/match statements. It can happen that this pattern recognition approach turns out to be too brittle: we might accidentally leave out cases when the pattern is s ...
````
Yep, we need to match the whole pattern, not just the Reduce. If some other random thing appears inside the pattern, then we have a problem. This worries me a bit as well...

For now, we'll try to make sure that we catch all variations of the pattern. We'll test this very carefully, with queries where window functions are in many different contexts (we'll definitely need fuzzing).

We'll also add a sentry logging for the situation where we detected the Reduce, but failed to detect the rest of the pattern. If this happens a few times in production (despite the careful testing with fuzzing), then we'll re-evaluate the decision of whether to have an explicit MIR representation for window functions. (Btw. adding a new MIR variant  might be easier later, when we have completed some optimizer refactorings.)

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,539 @@
+Some window functions are impossible to efficiently support in streaming, because sometimes small input changes cause big result changes. (E.g., if a new first element of a partition appears, then ROW_NUMBERs will change for the whole window partition.) So a realistic goal would be to support at least those cases where a small input change leads to a small output change.
+
+- LAG/LEAD (i.e., previous/next element of the window partition) (these don't have any frames)
+    - We aim for only offset 1 in the first version (i.e., the previous or the next element), which is the default. Bigger offsets not seen in user queries yet (i.e., when requesting to go back or forward by several rows).
```
Thanks, fixed!

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,539 @@
+        - SUMming an expression that is 0 for many rows.
+        - Window aggregations with an UNBOUNDED PRECEDING frame are fine if changes happen mostly at the end of partitions
+            - e.g., OVER an `ORDER BY time` if new elements are arriving typically with fresh timestamps. Such OVER clauses are popular in our [use case list](https://www.notion.so/Window-Functions-Use-Cases-6ad1846a7da942dc8fa28997d9c220dd).
+    - For frames encompassing the entire window partition (i.e., an UNBOUNDED frame and/or no ORDER BY), window aggregations can be simply translated to a standard grouped aggregation + a self-join. In case of these frames, small output changes often lead to big output changes, but similar exceptions exist as listed for "Large frames" above.
```
fixed

ggevay commented on doc/developer/design/20230110_window_functions.md:
````diff
@@ -0,0 +1,539 @@
+FROM transactions;
+```
+
+These operate on so-called **frames**, i.e., a certain subset of a window partition. Frames are specified in relation to the current row. For example, "sum up column `x` for the preceding 5 rows from the current row". For all the frame options, see https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS
````
Thanks, fixed this instance! I'll go through the whole document at the end to eliminate repetitions like this.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+This is not an unreasonable assumption, because a group is identified here by a value of the PARTITION BY expression + a value of the ORDER BY expression.
+Also note that if there is no ORDER BY, then groups might be large, but in this case we don’t employ Prefix Sum, but we transform away the window functions to grouped aggregation + self-join (as noted above).
+
+Alternatively, we could change Prefix Sum so that it correctly handles duplicate indexes.
```
Rewrote this one also. This will be handled by the hashing.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+##### 1.b. ROW_NUMBER, RANK, DENSE_RANK, PERCENT_RANK, CUME_DIST, NTILE
+
+For example: List the two biggest cities of each state:
+(Note that we can't directly write `ROW_NUMBER() <= 2`, because window functions are not allowed in WHERE clause.)
```
It's a limitation of SQL, just stemming from the order in which the different parts of the query are executed: FROM, WHERE, GROUP BY, HAVING, then window functions, then other things. Added an explanation.

Btw. Snowflake introduced the [QUALIFY keyword](https://docs.snowflake.com/en/sql-reference/constructs/qualify) as an elegant solution to this limitation.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+- We don’t handle such OVER clauses where the ORDER BY inside the OVER is on a String or other complex type. See a discussion on supported types below in the "Types" section.
+- In cases that we handle by Prefix Sum, the groups specified by the composite key of the PARTITION BY and the ORDER BY should be small, see the "Duplicate Indexes" section.
+
+# Details
```
I added a description of how prefix sum work internally, and various performance considerations.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+- Date/Time types are just a few integers. We’ll concatenate their bits.
+- I don’t know how to handle strings, so these are out of scope for now. (Not seen in user queries yet.)
+
+## Alternatives
```
Hmm, yes. I added a section that discusses an alternative rendering using a custom data structure.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+    - We aim for only offset 1 in the first version, which is the default. Bigger offsets not seen in user queries yet, but shouldn’t be a problem to add support later.
+    - IGNORE NULLS should be supported. (already seen in a user query) (easy)
+- Window aggregations
+    - Small frames: small output changes for small input changes.
```
Fixed

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,331 @@
+
+I built [a small prototype outside Materialize](https://github.com/ggevay/window-funcs), where I verified that the output values are correct, and that output is quickly updated for small input changes.
+
+For LAG/LEAD with *k > 1* (which computes the given expression not for the previous record, but for the record that was *k* records ago), the sum function could simply remember the last *k* values, acting on a `Vec<Val>` of length at most *k*, which would generalize `Option<Val>`. This works kind of ok for small *k*. A more complicated but probably better solution is to find the index for that el ...
```
I will clarify this.

vmarcos commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+
+Just by looking at the bit patterns in the queries, it computes the set of intervals (`requests`) from which it will put together the final sums. Many precomputed intervals might participate in the final sum calculation for more than one query, and therefore `requests` are deduplicated.
+
+For a general prefix sum, it can happen that some intervals in `requests` don't occur among the pre-aggregated intervals, meaning that the sum for these intervals is 0. Therefore, there is an outer join to add these 0s. However, this case cannot occur when we use prefix sum for window functions, because our `queries` and the original set of indexes from which `aggregate` started are actually the  ...
```
This paragraph suggests a change to DD's implementation of `broadcast`. I am wondering here if we are at the end going to use a Materialize-rolled copy of DD's prefix sum implementation with several changes or if we are planning to reuse DD's implementation directly in all (some?) cases? Perhaps we could be a bit more explicit regarding implementation strategy, either in this or another section?

ggevay replied:
We'll definitely need a Materialize-rolled copy. I added a sentence in the first paragraph of the `### Prefix Sum` section to make this clear.

vmarcos commented on doc/developer/design/20230110_window_functions.md:
````diff
@@ -0,0 +1,592 @@
+FROM cities;
+```
+
+For LAG/LEAD with an offset of 1, the sum function will just remember the previous value if it exists, and `None` if it does not. (And we can have a similar one for LEAD.) This has the following properties:
````
```suggestion
For LAG with an offset of 1, the sum function will just remember the previous value if it exists, and `None` if it does not. (And we can have a similar one for LEAD.) This has the following properties:
```

ggevay replied:
fixed, thx

vmarcos commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+    - Btw. we might translate some inequality self-joins to this one!
+- GROUPS: (One could say that this is probably not so often used, so no need to initially support it. However, the problem is that the solution for ROWS will probably build on this, and that is the default, so that one is often used.) We have to somehow translate the offset to a difference in the “index” of the prefix sum:
+    - Naive solution: Since this is a difference in DENSE_RANK between the current row and the ends of the interval, we could compute a DENSE_RANK, and then join with that to find the “index” (the value of the ORDER BY column). However, this would be slow, because DENSE_RANK is inherently not well incrementalizable: Small input changes can lead to lots of changes in the ranks.
+    - A better solution is to calculate a count aggregation on all ranges (with prefix sum’s `aggregate`) (let's call this `counts`), and then do a logarithmic search for the index with a nested `iterate`: We start with `(index, offset)` pairs for all the possible current elements in parallel, where the pair means that we need to move `index` down by `offset` (down when looking for the lower end  ...
```
This seems to be an important algorithmic component of the solution. Should we just include a small algorithm fragment here to go along with the description? I often found myself here trying to reconstruct this fragment in my mind when reading things like "a nested `iterate`".

ggevay replied:
I will definitely make this part more detailed.

vmarcos commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+
+These also operate based on a **frame**, similarly to window aggregations. (The above example query doesn't specify a frame, therefore it uses the default frame: from the beginning of the partition to the current row) They can be similarly implemented to window aggregations, i.e., we could “sum” up the relevant interval (that is not necessarily a prefix) with an appropriate sum function.
+
+Alternatively, we could make these a bit faster (except for NTH_VALUE) if we just find the index of the relevant end of the interval (i.e., left end for FIRST_VALUE), and then self-join.
```
Are we talking here about an alternative that will go into the Alternatives section or is this an optimization that is part of the proposal that you'd like to implement?

I was a bit confused here at what level you create the pattern with a self-join. Is it when you recognize different cases prior to rendering or is this special-cased when you perform rendering itself?

ggevay replied:
This is an optimization that is part of the proposal that I'd like to implement. I changed the wording to make this more clear.

The self-join is created in the MIR-to-LIR lowering, similarly to other prefix sum decisions. This is because finding the end of the frame interval is not expressible in MIR, as it is the same operation as finding the ends of frames for window aggregations.

vmarcos commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+
+Alternatively, we could make these a bit faster (except for NTH_VALUE) if we just find the index of the relevant end of the interval (i.e., left end for FIRST_VALUE), and then self-join.
+
+(And there are some special cases when we can transform away the window function usage: FIRST_VALUE with UNBOUNDED PRECEDING and LAST_VALUE with UNBOUNDED FOLLOWING should be transformed to just a (non-windowed) grouped aggregation + self-join instead of prefix sum trickery. Also, similarly for the case when there is no ORDER BY.)
```
This sounds like a paragraph, not a parenthesis. :-)

ggevay replied:
I removed the parenthesis, and extended the text a bit :)

vmarcos reviewed (commented):
Thanks for the very detailed write-up! I left a few comments where some clarifications / examples might be useful.

ggevay commented on doc/developer/design/20230110_window_functions.md:
````diff
@@ -0,0 +1,592 @@
+FROM measurements;
+```
+
+Certain window functions operate on a _window frame_, which is a subset of a partition. The default frame includes the rows from the first row of the partition up to the current row (or more accurately, to the last row of the peer group of the current row, where a peer group is a set of rows that are equal on both the `PARTITION BY` and the `ORDER BY`). For example, all aggregation functions can  ...
````
Yes, thanks, I've added a paragraph for peer groups before this paragraph.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+
+Note that this query doesn't compute just one value for each partition. Instead, it calculates a value for each input row: the sum of the same sensor's measurements that happened no later than the current input row.
+
+We can also explicitly specify a frame, i.e., how far it extends from the current row, both backwards and forwards. One option is to say `UNBOUNDED PRECEDING` or `UNBOUNDED FOLLOWING`, meaning that the frame extends to the beginning or end of the current partition. Another option is to specify an offset. For example, the following query computes a moving average (e.g., to have a smoother curve wh ...
```
I added a parenthesis to say that the frame can extend only within the same partition.

ggevay commented on doc/developer/design/20230110_window_functions.md:
```diff
@@ -0,0 +1,592 @@
+
+We'll use several approaches to solve the many cases mentioned in [Goals](#goals):
+
+1. We'll use [DD's prefix_sum](https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/algorithms/prefix_sum.rs) with some tricky sum functions and some generalizations.
```
Thanks! I added a paragraph at the end of the `#### Properties of DD's prefix sum implementation` section.

ggevay commented on doc/developer/design/20230110_window_functions.md:
````diff
@@ -0,0 +1,592 @@
+        RANGE BETWEEN '5 minutes' PRECEDING AND CURRENT ROW)
+FROM measurements;
+```
+
````
It's good that you asked this, because I realized I can't even think of a single good example... Any examples that I found have the problem that they either don't make much sense, or can be easily solved without the GROUPS frame mode. (There are a several not-so-good examples here: https://mjk.space/advances-sql-window-frames/) 

Maybe we can postpone supporting it until we see at least one actual user request.

vmarcos replied:
This is more of a product-level decision, but either way, it might be good to be explicit in the document that we are not supporting this functionality. My understanding is that we do not support it right now: https://github.com/MaterializeInc/materialize/blob/928a91fa97bea88f49b843763edd9343cf6b6f89/src/expr/src/relation/func.rs#L1006-L1007
